client/src/margo_client.c

```diff
@@ -888,6 +889,102 @@ int invoke_client_mread_rpc(unifyfs_client* client,
     return ret;
 }
 
+/* invokes the client metaget rpc function */
+int invoke_client_node_local_extents_get_rpc(unifyfs_client* client,
+                                             int num_req,
+                                             extents_list_t* read_req,
+                                             size_t* extent_count,
+                                             unifyfs_client_index_t** extents)
+{
+    /* check that we have initialized margo */
+    if (NULL == client_rpc_context) {
+        return UNIFYFS_FAILURE;
+    }
+
+    size_t extents_size = num_req * sizeof(unifyfs_extent_t);
+    void* buffer = malloc(extents_size);
+    if (NULL == buffer) {
+        return ENOMEM;
+    }
+
+    unifyfs_extent_t* int_extents = (unifyfs_extent_t*)buffer;
+    extents_list_t* cur = read_req;
+    for (int i = 0; i < num_req; i++) {
+        unifyfs_extent_t* ext = int_extents + i;
+        ext->gfid = cur->value.gfid;
+        ext->offset = cur->value.file_pos;
+        ext->length = cur->value.length;
+        cur = cur->next;
+    }
+    /* get handle to rpc function */
+    hg_handle_t handle = create_handle(
+            client_rpc_context->rpcs.node_local_extents_get_id);
+
+    /* fill in input struct */
+    unifyfs_node_local_extents_get_in_t in;
+    hg_return_t hret = margo_bulk_create(client_rpc_context->mid,
+                                         1, &buffer, &extents_size,
+                                         HG_BULK_READ_ONLY, &in.bulk_data);
+    if (hret != HG_SUCCESS) {
+        return UNIFYFS_ERROR_MARGO;
+    }
+    in.app_id = (int32_t) client->state.app_id;
+    in.client_id = (int32_t) client->state.client_id;
+    in.num_req = num_req;
+    in.bulk_size = extents_size;
+
+    /* call rpc function */
+    LOGDBG("invoking the node_local_extents_get rpc function in client");
+    double timeout = client_rpc_context->timeout;
+    int rc = forward_to_server(handle, &in, timeout);
+    if (rc != UNIFYFS_SUCCESS) {
+        LOGERR("forward of metaget rpc to server failed");
+        margo_destroy(handle);
+        return rc;
+    }
+
+    /* decode response */
+    int ret;
+    unifyfs_node_local_extents_get_out_t out;
+    hret = margo_get_output(handle, &out);
+    if (hret == HG_SUCCESS) {
+        LOGDBG("Got response ret=%" PRIi32, out.ret);
+        ret = (int) out.ret;
+        if (ret == (int) UNIFYFS_SUCCESS) {
+            *extent_count = out.extent_count;
+
+            void* out_buffer = pull_margo_bulk_buffer(handle, out.bulk_data,
+                                                  out.bulk_size, NULL);
+            *extents = calloc(out.extent_count, sizeof(unifyfs_client_index_t));
+            unifyfs_client_index_t* out_extents  =
+                    (unifyfs_client_index_t*) out_buffer;
+            for (int i = 0; i < out.extent_count; ++i) {
```

Comment:
why do you need this extra array allocation and copy for loop? just use 
`*extents = (unifyfs_client_index_t*) out_buffer;`

Reply:
Done